http/client.go

```diff
@@ -0,0 +1,412 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package http
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	blockstore "github.com/ipfs/boxo/blockstore"
+	sse "github.com/vito/go-sse/sse"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/events"
+)
+
+var _ client.DB = (*Client)(nil)
+
+// Client implements the client.DB interface over HTTP.
+type Client struct {
+	http *httpClient
+}
+
+func NewClient(rawURL string) (*Client, error) {
+	baseURL, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
+	httpClient := newHttpClient(baseURL.JoinPath("/api/v0"))
+	return &Client{httpClient}, nil
+}
+
+func (c *Client) NewTxn(ctx context.Context, readOnly bool) (datastore.Txn, error) {
+	query := url.Values{}
+	if readOnly {
+		query.Add("read_only", "true")
+	}
+
+	methodURL := c.http.baseURL.JoinPath("tx")
+	methodURL.RawQuery = query.Encode()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, methodURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	var txRes CreateTxResponse
+	if err := c.http.requestJson(req, &txRes); err != nil {
+		return nil, err
+	}
+	return &Transaction{txRes.ID, c.http}, nil
+}
+
+func (c *Client) NewConcurrentTxn(ctx context.Context, readOnly bool) (datastore.Txn, error) {
+	query := url.Values{}
+	if readOnly {
+		query.Add("read_only", "true")
+	}
+
+	methodURL := c.http.baseURL.JoinPath("tx", "concurrent")
+	methodURL.RawQuery = query.Encode()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, methodURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	var txRes CreateTxResponse
+	if err := c.http.requestJson(req, &txRes); err != nil {
+		return nil, err
+	}
+	return &Transaction{txRes.ID, c.http}, nil
+}
+
+func (c *Client) WithTxn(tx datastore.Txn) client.Store {
+	client := c.http.withTxn(tx.ID())
```

Comment:
thought: I'm just highlighting here that our current transaction ID implementation is a bit of a hack, and exposing it more here increases the need to improve the way we manage these IDs.